src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioErrorReportingService.ExceptionFormatting.cs

```diff
@@ -0,0 +1,172 @@
+        }
+
+        private static string GetAsyncStackTrace(Exception exception)
+        {
+            var stackTrace = new StackTrace(exception);
+
+            var stackFrames = stackTrace.GetFrames();
+            if (stackFrames == null)
+            {
+                return string.Empty;
+            }
+
+            var firstFrame = true;
+            var stringBuilder = new StringBuilder();
+
+            foreach (var frame in stackFrames)
+            {
+                var method = frame.GetMethod();
+                var declaringType = method?.DeclaringType;
+                if (declaringType != null && typeof(INotifyCompletion).IsAssignableFrom(declaringType))
+                {
+                    continue;
+                }
+
+                if (firstFrame)
+                {
+                    firstFrame = false;
+                }
+                else
+                {
+                    stringBuilder.Append(Environment.NewLine);
+                }
+
+                stringBuilder.AppendFormat("   at ");
+                var isAsync = FormatMethodName(stringBuilder, declaringType);
+                if (!isAsync)
+                {
+                    stringBuilder.Append(method?.Name);
+                    var methodInfo = method as MethodInfo;
+                    if (methodInfo?.IsGenericMethod == true)
+                    {
+                        FormatGenericArguments(stringBuilder, methodInfo.GetGenericArguments());
+                    }
+                }
+                else if (declaringType?.IsGenericType == true)
+                {
+                    FormatGenericArguments(stringBuilder, declaringType.GetGenericArguments());
+                }
+
+                stringBuilder.Append("(");
+                if (isAsync)
+                {
+                    stringBuilder.Append(ServicesVSResources.Unknown_parameters);
+                }
+                else
+                {
+                    FormatParameters(stringBuilder, method);
+                }
+
+                stringBuilder.Append(")");
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        private static bool FormatMethodName(StringBuilder stringBuilder, Type declaringType)
+        {
+            if (declaringType == null)
+            {
+                return false;
+            }
+
+            var isAsync = false;
+            var fullName = declaringType.FullName.Replace('+', '.');
+            if (typeof(IAsyncStateMachine).GetTypeInfo().IsAssignableFrom(declaringType))
+            {
+                isAsync = true;
+                stringBuilder.Append("async ");
+                var start = fullName.LastIndexOf('<');
+                var end = fullName.LastIndexOf('>');
+                if (start >= 0 && end >= 0)
+                {
+                    stringBuilder.Append(fullName.Remove(start, 1).Substring(0, end - 1));
```

Comment:
does substring work properly here since you're manipulating the string first?  Should you not trim off the end bit first, then trim the beginning?

(also, i'm not sure i understand what this is trying to do in the first place)